File: cpp/lib/testcontainers/core/SyncExecResult.cpp
#include <rust/cxx.h>
#include <rust_tc_bridge/lib.h>

#include "testcontainers/core/SyncExecResult.hpp"

#include "details/BoxHelper.hpp"
#include "details/OptionHelper.hpp"
#include "details/VectorHelper.hpp"

namespace testcontainers {

SyncExecResult::SyncExecResult(RsSyncExecResult *result) noexcept
    : rimpl_(result, [](RsSyncExecResult *r) {
        ::rs_sync_exec_result_destroy(details::box_from_raw(r));
      }) {}

SyncExecResult::SyncExecResult(SyncExecResult &&other) noexcept = default;

SyncExecResult &SyncExecResult::operator=(SyncExecResult &&other) noexcept = default;

SyncExecResult::~SyncExecResult() noexcept = default;

std::optional<std::int64_t> SyncExecResult::exit_code() noexcept {
  return utils::vec_to_optional(::rs_sync_exec_result_exit_code_opt(*rimpl_));
}

std::vector<std::uint8_t> SyncExecResult::stdout_to_vec() noexcept {
  return utils::vec_to_vector(::rs_sync_exec_result_stdout_to_vec(*rimpl_));
}

std::vector<std::uint8_t> SyncExecResult::stderr_to_vec() noexcept {
  return utils::vec_to_vector(::rs_sync_exec_result_stderr_to_vec(*rimpl_));
}

std::string SyncExecResult::stdout_to_string() {
  const auto out = stdout_to_vec();
  return std::string(out.begin(), out.end());
}

std::string SyncExecResult::stderr_to_string() {
  const auto err = stderr_to_vec();
  return std::string(err.begin(), err.end());
}

bool SyncExecResult::is_valid() const noexcept { return static_cast<bool>(rimpl_); }

} // namespace testcontainers

File: rust/src/core/exec/sync_exec_result.rs
use testcontainers::core::SyncExecResult;

pub struct RsSyncExecResult {
    pub result: SyncExecResult,
}

pub fn rs_sync_exec_result_exit_code_opt(result: &mut RsSyncExecResult) -> Result<Vec<i64>, String> {
    result
        .result
        .exit_code()
        .map_err(|e| format!("Failed to get exit code: {}", e))
        .map(|opt| opt.into_iter().collect())
}

pub fn rs_sync_exec_result_stdout_to_vec(result: &mut RsSyncExecResult) -> Result<Vec<u8>, String> {
    result
        .result
        .stdout_to_vec()
        .map_err(|e| format!("Failed to get stdout: {}", e))
}

pub fn rs_sync_exec_result_stderr_to_vec(result: &mut RsSyncExecResult) -> Result<Vec<u8>, String> {
    result
        .result
        .stderr_to_vec()
        .map_err(|e| format!("Failed to get stderr: {}", e))
}

pub fn rs_sync_exec_result_destroy(result: Box<RsSyncExecResult>) {
    drop(result);
}

impl RsSyncExecResult {
    pub fn new(result: SyncExecResult) -> Self {
        Self { result }
    }
}
